File: app/src/main/java/com/example/bloold/buildp/map/RouteActivity.java
package com.example.bloold.buildp.map;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;
import android.widget.Toast;

import com.example.bloold.buildp.R;
import com.example.bloold.buildp.common.IntentHelper;
import com.example.bloold.buildp.ui.RouteMapActivity;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.concurrent.ExecutionException;

public class RouteActivity extends AppCompatActivity implements AdapterView.OnItemClickListener {
    private TextView tvRoute;
    private AutoCompleteTextView atvOrigin;
    private AutoCompleteTextView atvDestination;
    private static final String LOG_TAG = "Google Places Autocomplete";
    private static final String PLACES_API_BASE = "https://maps.googleapis.com/maps/api/place";
    private static final String TYPE_AUTOCOMPLETE = "/autocomplete";
    private static final String OUT_JSON = "/json";
    private static final String API_KEY = "AIzaSyBsqnZcch_56fMZAFOeO_eSh2nsqqLFWGY";
    String mrPosition = null;
    String geoPosition = null;
    private static String mrTitle = null;
    private Toolbar toolbar;
    String PosCoods = null;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_route);
        toolbar = (Toolbar) findViewById(R.id.toolbar_actionbar);
        setSupportActionBar(toolbar);
        if(getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
            TextView tvTitle = toolbar.findViewById(R.id.tvTitle);
            tvTitle.setText("Построить маршрут");
            tvTitle.setVisibility(View.VISIBLE);
        }
       /* toolbar = (Toolbar) findViewById(R.id.toolbar);
        if(getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
             getSupportActionBar().setTitle("Маршрут");

        }*/
        Intent intentRoute = getIntent();
        mrPosition = intentRoute.getStringExtra("mrPosition");
        geoPosition = intentRoute.getStringExtra("geoPosition");
        mrTitle = intentRoute.getStringExtra("mrTitle");

        atvOrigin = (AutoCompleteTextView) findViewById(R.id.atvOrigin);
        atvDestination = (AutoCompleteTextView) findViewById(R.id.atvDestination);
        if(mrTitle!=null && mrPosition!=null ){
            atvDestination.setText(mrTitle);
        }
        if(geoPosition!=null && geoPosition!="00"){
            atvOrigin.setText("Мое местоположение");
        }

        tvRoute = (TextView) findViewById(R.id.tvRoute);
        tvRoute.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String origin = geoPosition;
                String destination = mrPosition;
                if (origin==null) {
                    Toast.makeText(RouteActivity.this, "Адрес пуст", Toast.LENGTH_SHORT).show();
                    return;
                }
                if (destination==null) {
                    Toast.makeText(RouteActivity.this, "Адрес пуст", Toast.LENGTH_SHORT).show();
                    return;
                }
                startActivity(new Intent(RouteActivity.this, RouteMapActivity.class)
                    .putExtra(IntentHelper.Companion.getEXTRA_ORIGIN(), origin)
                        .putExtra(IntentHelper.Companion.getEXTRA_DESTINATION(), destination));
            }
        });
        AutoCompleteTextView autoCompView = findViewById(R.id.atvOrigin);
        autoCompView.setAdapter(new RouteActivity.GooglePlacesAutocompleteAdapter(this, R.layout.list_item));
        autoCompView.setOnItemClickListener((parent, v, position, id) -> {
            String selection = (String)parent.getItemAtPosition(position);
            try {
                 new JSONGet().execute(selection).get();
                Thread.sleep(1000);
                 geoPosition = PosCoods;
            } catch (InterruptedException e) {
                e.printStackTrace();
            } catch (ExecutionException e) {
                e.printStackTrace();
            }

        });

        final AutoCompleteTextView autoCompView2 = (AutoCompleteTextView) findViewById(R.id.atvDestination);
        autoCompView2.setAdapter(new RouteActivity.GooglePlacesAutocompleteAdapter(this, R.layout.list_item));
        autoCompView2.setOnItemClickListener(new AdapterView.OnItemClickListener() {

            @Override
            public void onItemClick(AdapterView<?> parent, View v, int position,long id) {
                String selection = (String)parent.getItemAtPosition(position);

                try {
                    new JSONGet().execute(selection).get();
                    Thread.sleep(1000);
                    mrPosition = PosCoods;
                } catch (InterruptedException e) {
                    e.printStackTrace();
                } catch (ExecutionException e) {
                    e.printStackTrace();
                }

            }
        });

    }
    public class JSONGet extends AsyncTask<String, String, String>{
        @Override
        protected String doInBackground(String... params) {
            HttpURLConnection connection = null;
            BufferedReader reader = null;
            String result ="";
            try {
                String input = params[0];
                URL url = new URL("http://ruinnet.idefa.ru/api_app/object/index/?select[]=LAT&select[]=LNG&select[]=NAME&filter[0][?NAME]=" + URLEncoder.encode(input, "utf8"));

                connection = (HttpURLConnection) url.openConnection();
                connection.setRequestMethod("GET");
                connection.setRequestProperty("Device-Id","0000");
                String header = "Basic " + new String(android.util.Base64.encode("defa:defa".getBytes(), android.util.Base64.NO_WRAP));
                connection.addRequestProperty("Authorization", header);
                connection.connect();
                InputStream stream = connection.getInputStream();
                reader = new BufferedReader(new InputStreamReader(stream));

                String line ="";
                StringBuilder responseOutput = new StringBuilder();
                while((line = reader.readLine()) != null ) {
                    responseOutput.append(line);
                }
                result = responseOutput.toString();
                String finalJson = result.toString();
                JSONObject response = null;
                try {
                    response = new JSONObject(result);

                } catch (JSONException e) {
                    e.printStackTrace();
                }
                if (response !=null) {

                    JSONObject jsonObj = null;
                    try {
                        jsonObj = new JSONObject(result.toString());
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }

                    JSONObject DATA = null;
                    try {
                        DATA = new JSONObject(jsonObj.getString("DATA"));
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                    JSONArray predsJsonArray = null;
                    try {
                        predsJsonArray = DATA.getJSONArray("ITEMS");
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                    try {
                        PosCoods = predsJsonArray.getJSONObject(0).getString("LAT") + "," + predsJsonArray.getJSONObject(0).getString("LNG");
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }

            }catch (MalformedURLException e){
                e.printStackTrace();
            }catch(IOException e){
                e.printStackTrace();
            } finally {
                if (connection != null) {
                    connection.disconnect();
                }
                try {
                    if (reader != null) {
                        reader.close();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return  null;

        }

    }

   /* @SuppressLint("LongLogTag")
    public static ArrayList autocomplete(String input) {
        ArrayList resultList = null;

        HttpURLConnection conn = null;
        StringBuilder jsonResults = new StringBuilder();
        try {
            StringBuilder sb = new StringBuilder(PLACES_API_BASE + TYPE_AUTOCOMPLETE + OUT_JSON);
            sb.append("?key=" + API_KEY);
            sb.append("&input=" + URLEncoder.encode(input, "utf8"));

            URL url = new URL(sb.toString());
            conn = (HttpURLConnection) url.openConnection();
            InputStreamReader in = new InputStreamReader(conn.getInputStream());

            // Load the results into a StringBuilder
            int read;
            char[] buff = new char[1024];
            while ((read = in.read(buff)) != -1) {
                jsonResults.append(buff, 0, read);
            }
        } catch (MalformedURLException e) {
            Log.e(LOG_TAG, "Error processing Places API URL", e);
            return resultList;
        } catch (IOException e) {
            Log.e(LOG_TAG, "Error connecting to Places API", e);
            return resultList;
        } finally {
            if (conn != null) {
                conn.disconnect();
            }
        }

        try {
            // Create a JSON object hierarchy from the results
            JSONObject jsonObj = new JSONObject(jsonResults.toString());
            JSONArray predsJsonArray = jsonObj.getJSONArray("predictions");

            // Extract the Place descriptions from the results
            resultList = new ArrayList(predsJsonArray.length());
            for (int i = 0; i < predsJsonArray.length(); i++) {
                System.out.println(predsJsonArray.getJSONObject(i).getString("description"));
                System.out.println("============================================================");
                resultList.add(predsJsonArray.getJSONObject(i).getString("description"));
            }
        } catch (JSONException e) {
            Log.e(LOG_TAG, "Cannot process JSON results", e);
        }

        return resultList;
    }*/
   @SuppressLint("LongLogTag")
   public static ArrayList autocomplete(String input) {
       ArrayList resultList = null;

       HttpURLConnection conn = null;
       StringBuilder jsonResults = new StringBuilder();
       try {
           StringBuilder sb = new StringBuilder("http://ruinnet.idefa.ru/api_app/object/index/?select[]=NAME&filter[0][?NAME]=" + URLEncoder.encode(input, "utf8"));

           URL url = new URL(sb.toString());
           conn = (HttpURLConnection) url.openConnection();
           InputStreamReader in = new InputStreamReader(conn.getInputStream());

           // Load the results into a StringBuilder
           int read;
           char[] buff = new char[1024];
           while ((read = in.read(buff)) != -1) {
               jsonResults.append(buff, 0, read);
           }
       } catch (MalformedURLException e) {
           Log.e(LOG_TAG, "Error processing Places API URL", e);
           return resultList;
       } catch (IOException e) {
           Log.e(LOG_TAG, "Error connecting to Places API", e);
           return resultList;
       } finally {
           if (conn != null) {
               conn.disconnect();
           }
       }

       try {
           // Create a JSON object hierarchy from the results
           JSONObject jsonObj = new JSONObject(jsonResults.toString());

           JSONObject DATA = new JSONObject(jsonObj.getString("DATA"));
           JSONArray predsJsonArray = DATA.getJSONArray("ITEMS");

           // Extract the Place descriptions from the results
           resultList = new ArrayList(predsJsonArray.length());
           for (int i = 0; i < predsJsonArray.length(); i++) {
               System.out.println(predsJsonArray.getJSONObject(i).getString("NAME"));
               System.out.println("============================================================");
               resultList.add(predsJsonArray.getJSONObject(i).getString("NAME"));
           }
       } catch (JSONException e) {
           Log.e(LOG_TAG, "Cannot process JSON results", e);
       }

       return resultList;
   }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {

    }

    class GooglePlacesAutocompleteAdapter extends ArrayAdapter implements Filterable {
        private ArrayList resultList;

        public GooglePlacesAutocompleteAdapter(Context context, int textViewResourceId) {
            super(context, textViewResourceId);
        }

        @Override
        public int getCount() {
            return resultList.size();
        }

        @Override
        public String getItem(int index) {
            return (String) resultList.get(index);
        }

        @Override
        public Filter getFilter() {
            Filter filter = new Filter() {
                @Override
                protected FilterResults performFiltering(CharSequence constraint) {
                    FilterResults filterResults = new FilterResults();
                    if (constraint != null) {
                        // Retrieve the autocomplete results.
                        resultList = autocomplete(constraint.toString());

                        // Assign the data to the FilterResults
                        filterResults.values = resultList;
                        filterResults.count = resultList.size();
                    }
                    return filterResults;
                }

                @Override
                protected void publishResults(CharSequence constraint, FilterResults results) {
                    if (results != null && results.count > 0) {
                        notifyDataSetChanged();
                    } else {
                        notifyDataSetInvalidated();
                    }
                }
            };
            return filter;
        }
    }
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        //noinspection SimplifiableIfStatement
        if (id == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/com/example/bloold/buildp/ui/RouteActivity.kt
package com.example.bloold.buildp.ui

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.databinding.DataBindingUtil
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AlertDialog
import android.view.MenuItem
import android.view.View
import android.widget.*
import com.example.bloold.buildp.R
import com.example.bloold.buildp.api.ServiceGenerator
import com.example.bloold.buildp.api.data.BaseResponseWithDataObject
import com.example.bloold.buildp.common.IntentHelper
import com.example.bloold.buildp.common.RxHelper
import com.example.bloold.buildp.components.NetworkActivity
import com.example.bloold.buildp.databinding.ActivityRouteBinding
import com.example.bloold.buildp.model.MyItem
import io.reactivex.observers.DisposableSingleObserver
import java.net.*

class RouteActivity : NetworkActivity()
{
    private lateinit var mBinding:ActivityRouteBinding

    //----- Получение координа пользователя ------
    private val PERMISSIONS_LOCATION = 234
    private var userLocation: Location? = null
    private var locationListener: LocationListener? = null
    //Получили актуальные координаты пользователя(не последние)
    private var gotActualLocation = false

    private var origin:MyItem?=null
    private var destination:MyItem?=null

    private var itemObject:MyItem?=null

    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_route)
        mBinding.listener=this
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        mBinding.atvOrigin.setAdapter<RouteActivity.GooglePlacesAutocompleteAdapter>(
                RouteActivity.GooglePlacesAutocompleteAdapter(this, android.R.layout.simple_dropdown_item_1line))
        mBinding.atvOrigin.setOnItemClickListener { parent, v, position, id ->
            val selection = parent.getItemAtPosition(position) as MyItem
            mBinding.atvOrigin.setText(selection.snippet)
            origin=selection
        }

        mBinding.atvDestination.onItemClickListener = AdapterView.OnItemClickListener { parent, v, position, id ->
            val selection = parent.getItemAtPosition(position) as MyItem
            mBinding.atvDestination.setText(selection.snippet)
            destination=selection
        }

        loadObject(intent.getIntExtra(IntentHelper.EXTRA_OBJECT_ID, -1))
    }

    override fun onResume() {
        super.onResume()
        checkGPSPermissions()
    }
    override fun onPause() {
        super.onPause()
        stopLocationUpdates()
    }
    private fun checkGPSPermissions() {
        // Here, thisActivity is the current activity
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED || ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    arrayOf(android.Manifest.permission.ACCESS_COARSE_LOCATION, android.Manifest.permission.ACCESS_FINE_LOCATION),
                    PERMISSIONS_LOCATION)
        } else
            startGettingUserLocation()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            PERMISSIONS_LOCATION -> {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    startGettingUserLocation()
                } else {
                    AlertDialog.Builder(this)
                            .setMessage(R.string.need_gps_permission)
                            .setCancelable(false)
                            .setPositiveButton(android.R.string.ok) { _, _ -> checkGPSPermissions() }
                            .setNegativeButton(android.R.string.cancel) { _, _ -> this.finish() }.show()
                }
                return
            }
        }// other 'case' lines to check for other
        // permissions this app might request
    }
    private fun startGettingUserLocation() {
        //Пока определяются текущие координаты, мы получаем посление
        if (userLocation == null) {
            userLocation = fetchLastLocation()
        }
        if (!gotActualLocation) fetchActualUserLocation()//Запускаем получение актуальных координат
    }
    // -------- Получение местоположения пользователя ----------
    private fun fetchLastLocation(): Location? {
        try {
            val locationManager = this.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            val lastLocation = locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER)
            return lastLocation
        } catch (ex: SecurityException) {
            ex.printStackTrace()
            Toast.makeText(this, R.string.gps_not_enabled, Toast.LENGTH_LONG).show()
        }

        return null
    }
    private fun fetchActualUserLocation() {
        // Define a listener that responds to location updates
        if (locationListener == null) {
            locationListener = object : LocationListener {
                override fun onLocationChanged(location: Location) {
                    // Called when a new location is found by the network location provider.
                    userLocation = location
                    stopLocationUpdates()
                    gotActualLocation = true
                    locationListener = null
                }

                override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {}

                override fun onProviderEnabled(provider: String) {}

                override fun onProviderDisabled(provider: String) {}
            }
        }

        try {
            val locationManager = this.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 0, 0f, locationListener)
        } catch (ex: SecurityException) {
            ex.printStackTrace()
            Toast.makeText(this, R.string.gps_not_enabled, Toast.LENGTH_LONG).show()
        }
    }
    private fun stopLocationUpdates() {
        if (locationListener != null) {
            val locationManager = this.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            locationManager.removeUpdates(locationListener)
        }
    }

    fun onRouteClick(v:View)
    {
        if (userLocation == null&&origin==null) {
            Toast.makeText(this@RouteActivity, R.string.need_set_origin_location, Toast.LENGTH_SHORT).show()
            return
        }
        if (itemObject == null&&destination==null) {
            Toast.makeText(this@RouteActivity, R.string.need_set_destination_location, Toast.LENGTH_SHORT).show()
            return
        }
        RouteMapActivity.launch(this, origin?.lat?:userLocation?.latitude?:0.0, origin?.lng?:userLocation?.longitude?:0.0,
                destination?.lat?:itemObject?.lat?:0.0, destination?.lng?:itemObject?.lng?:0.0)
    }
    private fun showProgress(showProgress: Boolean) {
        mBinding.flLoading.visibility = if (showProgress) View.VISIBLE else View.GONE
    }
    private fun loadObject(objectId:Int)
    {
        compositeDisposable.add(ServiceGenerator.serverApi.searchMapObjects(HashMap<String,String>().apply { put("filter[OBJECT_ID]",objectId.toString()) })
                .compose(RxHelper.applySchedulers())
                .doOnSubscribe { showProgress(true) }
                .doFinally {
                    showProgress(false)
                    if(itemObject==null)
                    {
                        Toast.makeText(applicationContext, R.string.object_not_found, Toast.LENGTH_SHORT).show()
                        finish()
                    }
                    else updateUI()
                }
                .subscribeWith(object : DisposableSingleObserver<BaseResponseWithDataObject<MyItem>>() {
                    override fun onSuccess(result: BaseResponseWithDataObject<MyItem>) {
                        itemObject=result.data?.items?.firstOrNull()
                    }
                    override fun onError(e: Throwable) {
                        e.printStackTrace()
                        if (e is UnknownHostException || e is ConnectException)
                            Toast.makeText(this@RouteActivity, R.string.error_check_internet, Toast.LENGTH_SHORT).show()
                        else
                            Toast.makeText(this@RouteActivity, R.string.server_error, Toast.LENGTH_SHORT).show()
                    }
                }))
    }
    private fun updateUI()
    {
        itemObject?.let {
            mBinding.atvDestination.setText(it.snippet)
            mBinding.atvDestination.setAdapter<RouteActivity.GooglePlacesAutocompleteAdapter>(
                    RouteActivity.GooglePlacesAutocompleteAdapter(this, android.R.layout.simple_dropdown_item_1line))
        }
    }

    internal class GooglePlacesAutocompleteAdapter(context: Context, textViewResourceId: Int) : ArrayAdapter<MyItem>(context, textViewResourceId), Filterable {
        private var resultList: List<MyItem>? = null

        override fun getCount(): Int {
            return resultList!!.size
        }

        override fun getItem(index: Int): MyItem? {
            return resultList!![index]
        }

        override fun getFilter(): Filter {
            return object : Filter() {
                override fun performFiltering(constraint: CharSequence?): FilterResults {
                    val filterResults = FilterResults()
                    if (constraint != null) {
                        // Retrieve the autocomplete results.
                        resultList = autocomplete(constraint.toString())

                        // Assign the data to the FilterResults
                        filterResults.values = resultList
                        filterResults.count = resultList!!.size
                    }
                    return filterResults
                }

                override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
                    if (results != null && results.count > 0) {
                        notifyDataSetChanged()
                    } else {
                        notifyDataSetInvalidated()
                    }
                }
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId

        if (id == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        fun launch(cntx: Activity, objectId: Int) {
            cntx.startActivity(Intent(cntx, RouteActivity::class.java)
                    .putExtra(IntentHelper.EXTRA_OBJECT_ID, objectId))
        }
        @SuppressLint("LongLogTag")
        fun autocomplete(input: String): List<MyItem>? {
            return ServiceGenerator.serverApi.searchMapObjectsByName(input).blockingGet()?.data?.items?.toList()
        }
    }

}
